### ISSUE TITLE
Cocos2D needs to support all 'supportedInterfaceOrientations' options

### ISSUE THREAD

**Issue by Ben-G** [cocos2d/cocos2d-objc#891]
Currently the supported interface orientations are determined by the `CCNavigationController` inside `CCAppDelegate`:

```
-(NSUInteger)supportedInterfaceOrientations
{
    if ([_screenOrientation isEqual:CCScreenOrientationAll])
    {
        return UIInterfaceOrientationMaskAll;
    }
    else if ([_screenOrientation isEqual:CCScreenOrientationPortrait])
    {
        return UIInterfaceOrientationMaskPortrait | UIInterfaceOrientationMaskPortraitUpsideDown;
    }
    else
    {
        return UIInterfaceOrientationMaskLandscape;
    }
}
```

The supported screen orientation is determined by the `screenOrientation` property of `CCNavigationController` which only has three valid values:

```
NSString* const CCScreenOrientationLandscape = @"CCScreenOrientationLandscape";
NSString* const CCScreenOrientationPortrait = @"CCScreenOrientationPortrait";
NSString* const CCScreenOrientationAll = @"CCScreenOrientationAll"
```

If games want to restrict the supported device orientation to only one orientation (e.g. LandscapeLeft) the Cocos2D source code needs to be modified to change the implementation of the `supportedInterfaceOrientations` method.

`CCNavigationController` should expose all possible supported device orientations, either by adding more string constants or by another solution.

**Comment by Metalboyblue** [comment]
Agreed.  I am having this problem now.  Also, I would like to add the request to set this parameter on a "Per-Scene" basis.  I have a unique situation I am sure but some of the scenes work best in landscape and some in portrait.

**Comment by slembcke** [comment]
You could override the method easily enough without needing to change Cocos code. I'm a bit hesitant to add too many options or configurability to CCAppDelegate since it's purpose was to simplify the app delegate experience from 2.x. Using CCAppDelegate is optional, it's just meant to be a convenience that does what most people want with minimal fuss.

As for per-scene transitions, you could override the supportedInterfaceOrientations and make it more dynamic. It would get much more tricky if you wanted to have transitions involved though.

**Comment by Ben-G** [comment]
At the moment the methods are implemented inside of `CCNavigationController` which is instantiated inside of `setupCocos2dWithOptions`. So I would have to override `setupCocos2dWithOptions` which is the majority of the code in `CCAppDelegate` in order to instantiate my subclass of `CCNavigationController`. Or am I missing something?

To me it seems that adding the additional options for the orientations would not add a lot of complexity to `CCAppDelegate`.

**Comment by slembcke** [comment]
You are of course correct. Shoot. I remember thinking it was way easier than that.

Does it make sense to expose this using more options for the setup dictionary? (Since we didn't use bitmasks... urk!) Should the navigation controller forward the method call to the app delegate so that it can be overridden? That would make more dynamic usages possible too.

**Comment by Ben-G** [comment]
Yes, I was thinking of the second option and I think I prefer that one. It is also compatible with SpriteBuilder projects, because SpriteBuilder would otherwise override the setup .plist file.

**Comment by iMobCoding** [comment]
Here is a pull request that solves this https://github.com/cocos2d/cocos2d-iphone/pull/930
Basically what @slembcke suggested as 2nd option.
